contrib/pyzfs/libzfs_core/test/test_libzfs_core.py

```diff
@@ -0,0 +1,3730 @@
+        bad_fd = hard + 1
+        with self.assertRaises(lzc_exc.StreamIOError) as ctx:
+            lzc.lzc_send(snap, None, bad_fd)
+        self.assertEquals(ctx.exception.errno, errno.EBADF)
+
+    def test_send_to_broken_pipe(self):
+        snap = ZFSTest.pool.makeName("fs1@snap")
+        lzc.lzc_snapshot([snap])
+
+        proc = subprocess.Popen(['true'], stdin=subprocess.PIPE)
+        proc.wait()
+        with self.assertRaises(lzc_exc.StreamIOError) as ctx:
+            lzc.lzc_send(snap, None, proc.stdin.fileno())
+        self.assertEquals(ctx.exception.errno, errno.EPIPE)
+
+    def test_send_to_broken_pipe_2(self):
+        snap = ZFSTest.pool.makeName("fs1@snap")
+        with zfs_mount(ZFSTest.pool.makeName("fs1")) as mntdir:
+            with tempfile.NamedTemporaryFile(dir=mntdir) as f:
+                for i in range(1024):
+                    f.write('x' * 1024)
+                f.flush()
+                lzc.lzc_snapshot([snap])
+
+        proc = subprocess.Popen(['sleep', '2'], stdin=subprocess.PIPE)
+        with self.assertRaises(lzc_exc.StreamIOError) as ctx:
+            lzc.lzc_send(snap, None, proc.stdin.fileno())
+        self.assertTrue(ctx.exception.errno == errno.EPIPE or
+                        ctx.exception.errno == errno.EINTR)
+
+    def test_send_to_ro_file(self):
+        snap = ZFSTest.pool.makeName("fs1@snap")
+        lzc.lzc_snapshot([snap])
+
+        with tempfile.NamedTemporaryFile(
+                suffix='.ztream', delete=False) as output:
+            # tempfile always opens a temporary file in read-write mode
+            # regardless of the specified mode, so we have to open it again.
+            os.chmod(output.name, stat.S_IRUSR)
+            fd = os.open(output.name, os.O_RDONLY)
+            with self.assertRaises(lzc_exc.StreamIOError) as ctx:
+                lzc.lzc_send(snap, None, fd)
+            os.close(fd)
+        self.assertEquals(ctx.exception.errno, errno.EBADF)
+
+    def test_recv_full(self):
+        src = ZFSTest.pool.makeName("fs1@snap")
+        dst = ZFSTest.pool.makeName("fs2/received-1@snap")
+
+        with temp_file_in_fs(ZFSTest.pool.makeName("fs1")) as name:
+            lzc.lzc_snapshot([src])
+
+        with tempfile.TemporaryFile(suffix='.ztream') as stream:
+            lzc.lzc_send(src, None, stream.fileno())
+            stream.seek(0)
+            lzc.lzc_receive(dst, stream.fileno())
+
+        name = os.path.basename(name)
+        with zfs_mount(src) as mnt1, zfs_mount(dst) as mnt2:
+            self.assertTrue(
+                filecmp.cmp(
+                    os.path.join(mnt1, name), os.path.join(mnt2, name), False))
+
+    def test_recv_incremental(self):
+        src1 = ZFSTest.pool.makeName("fs1@snap1")
+        src2 = ZFSTest.pool.makeName("fs1@snap2")
+        dst1 = ZFSTest.pool.makeName("fs2/received-2@snap1")
+        dst2 = ZFSTest.pool.makeName("fs2/received-2@snap2")
+
+        lzc
```

Comment:
The only remaining tests that are still failing ("test_recv_..._origin") seem to be broken by _Illumos 5925 zfs receive -o origin=_ (fcff0f35bd522076bdda7491c88a91cc0aa531a3):

```
root@linux:~# stap -g -DMAXSTRINGLEN=10240 -v -d zfs -e '
> probe
> module("zfs").function("zfs_ioc_recv*").call,
> module("zfs").function("recv_begin_check_existing_impl").call,
> module("zfs").function("dmu_recv_resume_begin_check").call,
> module("zfs").function("dmu_recv_begin_sync").call,
> module("zfs").function("dmu_recv_begin_check").call,
> module("zfs").function("dsl_sync_task").call,
> module("zfs").function("dmu_recv_begin").call
> { printf(" -> %s %s\n", ppfunc(), $$parms$$); }
> probe
> module("zfs").function("recv_begin_check_existing_impl").return,
> module("zfs").function("dmu_recv_resume_begin_check").return,
> module("zfs").function("dmu_recv_begin_sync").return,
> module("zfs").function("dmu_recv_begin_check").return,
> module("zfs").function("dsl_sync_task").return,
> module("zfs").function("dmu_recv_begin").return,
> module("zfs").function("zfs_ioc_recv*").return
> { printf(" <- %s %s\n", ppfunc(), $$return$$); }' -c "python -m unittest --verbose libzfs_core.test.test_libzfs_core.ZFSTest.test_recv_incremental_non_clone_but_set_origin"
...
 -> zfs_ioc_recv_impl tofs="pool.fc6cfadf-7ce8-45c4-9a0d-8c29cc94d0f5/fs2/received-20" tosnap="snap2" origin="pool.fc6cfadf-7ce8-45c4-9a0d-8c29cc94d0f5/fs2/received-20@snap1" ...
 -> dmu_recv_begin tofs="pool.fc6cfadf-7ce8-45c4-9a0d-8c29cc94d0f5/fs2/received-20" tosnap="snap2" drr_begin={... .drr_u={.drr_begin={.... .drr_flags=4, ..., .drr_toname="pool.fc6cfadf-7ce8-45c4-9a0d-8c29cc94d0f5/fs1@snap2"}, ...} force=0 resumable=0 origin="pool.fc6cfadf-7ce8-45c4-9a0d-8c29cc94d0f5/fs2/received-20@snap1", ...}
 -> dsl_sync_task pool="pool.fc6cfadf-7ce8-45c4-9a0d-8c29cc94d0f5/fs2/received-20" ...
 -> dmu_recv_begin_check ...
 <- dmu_recv_begin_check return=22
 <- dsl_sync_task return=22
 <- dmu_recv_begin return=22
 <- zfs_ioc_recv_impl return=22
 <- zfs_ioc_recv return=22
```

With `drr_flags=DRR_FLAG_FREERECORDS` and `origin="pool.fc6cfadf-7ce8-45c4-9a0d-8c29cc94d0f5/fs2/received-20@snap1"` we are failing here:

https://github.com/zfsonlinux/zfs/blob/e086e717c3fc2c5b0867ee4b5ec07200a53b8dac/module/zfs/dmu_send.c#L1687-L1695